File: examples/go/blink-button.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// blink-button.go — blink an LED and wait for a button using the sysfs GPIO
// interface, with the Go standard library only (no external modules).
//
// Edge detection uses select(2) with the value fd in the *exception* set — the way
// sysfs GPIO edges are surfaced to select(). Equivalent to poll()'s POLLPRI.
//
//	LED    = GPIO17 (output)
//	BUTTON = GPIO27 (input, rising edge)
//
// Run:  go run blink-button.go   (or: go build -o blink-button . && ./blink-button)
package main

import (
	"fmt"
	"os"
	"strings"
	"syscall"
	"time"
)

const (
	sysfs  = "/sys/class/gpio"
	led    = 17
	button = 27
)

func write(path, val string) error { return os.WriteFile(path, []byte(val), 0) }
func export(pin int)               { _ = write(sysfs+"/export", fmt.Sprint(pin)) }   // ignore EBUSY
func unexport(pin int)             { _ = write(sysfs+"/unexport", fmt.Sprint(pin)) }

func main() {
	// --- blink the LED ---
	export(led)
	if err := write(fmt.Sprintf("%s/gpio%d/direction", sysfs, led), "out"); err != nil {
		panic(err)
	}
	fmt.Printf("Blinking GPIO%d...\n", led)
	valPath := fmt.Sprintf("%s/gpio%d/value", sysfs, led)
	for i := 0; i < 6; i++ {
		level := "0"
		if i%2 == 0 {
			level = "1"
		}
		write(valPath, level)
		time.Sleep(100 * time.Millisecond)
	}
	write(valPath, "0")

	// --- set up the button: input, rising edge ---
	export(button)
	write(fmt.Sprintf("%s/gpio%d/direction", sysfs, button), "in")
	write(fmt.Sprintf("%s/gpio%d/edge", sysfs, button), "rising")

	f, err := os.Open(fmt.Sprintf("%s/gpio%d/value", sysfs, button))
	if err != nil {
		panic(err)
	}
	defer f.Close()
	buf := make([]byte, 8)
	f.Seek(0, 0)
	f.Read(buf) // initial read clears any pending state (rearm)

	fd := int(f.Fd())
	fmt.Printf("WAITING for button press on GPIO%d (up to 5s)...\n", button)

	var except syscall.FdSet
	except.Bits[fd/64] |= int64(1) << (uint(fd) % 64)
	tv := syscall.Timeval{Sec: 5, Usec: 0}

	n, err := syscall.Select(fd+1, nil, nil, &except, &tv)
	if err != nil && err != syscall.EINTR {
		panic(err)
	}

	rc := 3
	if n > 0 {
		f.Seek(0, 0)
		m, _ := f.Read(buf)
		fmt.Printf("Button pressed! value=%s\n", strings.TrimSpace(string(buf[:m])))
		rc = 0
	} else {
		fmt.Println("timeout: no button press")
	}

	unexport(led)
	unexport(button)
	os.Exit(rc)
}

File: examples/python-sysfs/blink-button.py
#!/usr/bin/env python3
# SPDX-License-Identifier: AGPL-3.0-or-later
"""blink-button.py — blink an LED and wait for a button using the sysfs GPIO
interface directly, with Python's standard library only.

Edge detection uses select.poll() with POLLPRI on gpio27/value — the same poll(2)
mechanism a real Raspberry Pi sysfs consumer uses. No third-party packages.

    LED    = GPIO17 (output)
    BUTTON = GPIO27 (input, rising edge)

Note: this deliberately does NOT use RPi.GPIO. RPi.GPIO drives the SoC registers via
/dev/gpiomem, not sysfs, so it cannot work against a sysfs simulator. Direct file I/O
(shown here) or python-periphery's sysfs mode both work.

Run:  python3 blink-button.py
"""
import select
import sys
import time

SYSFS = "/sys/class/gpio"
LED = 17
BUTTON = 27


def write(path, value):
    with open(path, "w") as f:
        f.write(str(value))


def export(pin):
    try:
        write(f"{SYSFS}/export", pin)
    except OSError:
        pass  # already exported


def unexport(pin):
    try:
        write(f"{SYSFS}/unexport", pin)
    except OSError:
        pass


def main():
    # --- blink the LED ---
    export(LED)
    write(f"{SYSFS}/gpio{LED}/direction", "out")
    print(f"Blinking GPIO{LED}...")
    for i in range(6):
        write(f"{SYSFS}/gpio{LED}/value", i % 2)
        time.sleep(0.1)
    write(f"{SYSFS}/gpio{LED}/value", 0)

    # --- set up the button: input, rising-edge interrupt ---
    export(BUTTON)
    write(f"{SYSFS}/gpio{BUTTON}/direction", "in")
    write(f"{SYSFS}/gpio{BUTTON}/edge", "rising")

    value = open(f"{SYSFS}/gpio{BUTTON}/value", "r")
    value.read()  # initial read clears any pending state (rearm)

    poller = select.poll()
    poller.register(value, select.POLLPRI | select.POLLERR)

    print(f"WAITING for button press on GPIO{BUTTON} (up to 5s)...", flush=True)
    events = poller.poll(5000)  # milliseconds

    if events:
        value.seek(0)
        level = value.read().strip()
        print(f"Button pressed! value={level}")
        rc = 0
    else:
        print("timeout: no button press")
        rc = 3

    value.close()
    unexport(LED)
    unexport(BUTTON)
    sys.exit(rc)


if __name__ == "__main__":
    main()
